### PR TITLE
Add endpoint listing GCP machine and disk types

### LINKED ISSUE #3659: Add an API endpoint to fetch the list of node sizes on GCP
**User Story**
As a user, I'd like to be able to pick one of the currently available machine sizes, instead of relying on the values currently hardcoded in the dashboard. This is analogous to how we already fetch machine sizes in other providers.

**Acceptance criteria**
- [ ] a new endpoint /providers/gcp/sizes exists that returns the list of available machine sizes

**See also**
- https://cloud.google.com/compute/docs/reference/rest/v1/machineTypes/list

### PR DESCRIPTION
**Pull request by maciaszczykm**
**What this PR does / why we need it**: Adds an endpoint listing GCP machine sizes. Filters out ones with memory below 2 GBs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: Fixes #3659 

**Special notes for your reviewer**: Example output to `GET http://127.0.0.1:8080/api/v1/providers/gcp/machinetypes`:

```json
[
    {
        "name": "n1-highcpu-16",
        "description": "16 vCPUs, 14.4 GB RAM",
        "memory": 14746
    },
    {
        "name": "n1-highcpu-32",
        "description": "32 vCPUs, 28.8 GB RAM",
        "memory": 29491
    },
    {
        "name": "n1-highcpu-4",
        "description": "4 vCPUs, 3.6 GB RAM",
        "memory": 3686
    },
    {
        "name": "n1-highcpu-64",
        "description": "64 vCPUs, 57.6 GB RAM",
        "memory": 58982
    },
    {
        "name": "n1-highcpu-8",
        "description": "8 vCPUs, 7.2 GB RAM",
        "memory": 7373
    },
    {
        "name": "n1-highcpu-96",
        "description": "96 vCPUs, 86 GB RAM",
        "memory": 88474
    },
    {
        "name": "n1-highmem-16",
        "description": "16 vCPUs, 104 GB RAM",
        "memory": 106496
    },
    {
        "name": "n1-highmem-2",
        "description": "2 vCPUs, 13 GB RAM",
        "memory": 13312
    },
    {
        "name": "n1-highmem-32",
        "description": "32 vCPUs, 208 GB RAM",
        "memory": 212992
    },
    {
        "name": "n1-highmem-4",
        "description": "4 vCPUs, 26 GB RAM",
        "memory": 26624
    },
    {
        "name": "n1-highmem-64",
        "description": "64 vCPUs, 416 GB RAM",
        "memory": 425984
    },
    {
        "name": "n1-highmem-8",
        "description": "8 vCPUs, 52 GB RAM",
        "memory": 53248
    },
    {
        "name": "n1-highmem-96",
        "description": "96 vCPUs, 624 GB RAM",
        "memory": 638976
    },
    {
        "name": "n1-standard-1",
        "description": "1 vCPU, 3.75 GB RAM",
        "memory": 3840
    },
    {
        "name": "n1-standard-16",
        "description": "16 vCPUs, 60 GB RAM",
        "memory": 61440
    },
    {
        "name": "n1-standard-2",
        "description": "2 vCPUs, 7.5 GB RAM",
        "memory": 7680
    },
    {
        "name": "n1-standard-32",
        "description": "32 vCPUs, 120 GB RAM",
        "memory": 122880
    },
    {
        "name": "n1-standard-4",
        "description": "4 vCPUs, 15 GB RAM",
        "memory": 15360
    },
    {
        "name": "n1-standard-64",
        "description": "64 vCPUs, 240 GB RAM",
        "memory": 245760
    },
    {
        "name": "n1-standard-8",
        "description": "8 vCPUs, 30 GB RAM",
        "memory": 30720
    },
    {
        "name": "n1-standard-96",
        "description": "96 vCPUs, 360 GB RAM",
        "memory": 368640
    }
]
```

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: api/pkg/api/v1/types.go [hunk 1]
```diff
@@ -127,6 +127,18 @@ type Datacenter struct {
 	Seed     bool             `json:"seed,omitempty"`
 }
 
+// GCPMachineTypeList represents an array of GCP machine types.
+// swagger:model GCPMachineTypeList
+type GCPMachineTypeList []GCPMachineType
+
+// GCPMachineType represents a object of GCP machine type.
+// swagger:model GCPMachineType
+type GCPMachineType struct {
```
Comment (zreigz) [inline api/pkg/api/v1/types.go]: can you change it to GCPMachineSize like for other providers: https://github.com/kubermatic/kubermatic/blob/22b1789d4bd6d0c3129c5c6b8833d96c763e2e15/api/pkg/api/v1/types.go#L169

File: api/pkg/api/v1/types.go [hunk 2]
```diff
@@ -127,6 +127,18 @@ type Datacenter struct {
...
+	Name        string `json:"name"`
+	Description string `json:"description"`
```
Comment (zreigz) [inline api/pkg/api/v1/types.go:149]: can you replace description for 
`VCPUs        int      `json:"vcpus"`` 

to have only resources like for other providers

**Comment by maciaszczykm** [comment]
/hold

I will also add fetch for disk types here.

File: api/pkg/handler/routes_v1.go [hunk 3]
```diff
@@ -420,6 +424,28 @@ func (r Routing) listCredentials() http.Handler {
...
+//
+// Lists machine types from GCP
+//
+//     Produces:
+//     - application/json
+//
+//     Responses:
+//       default: errorResponse
+//       200: GCPMachineTypeList
+func (r Routing) listGCPMachineTypes() http.Handler {
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: can you change for `listGCPSizes`

File: api/pkg/handler/routes_v1.go [hunk 4]
```diff
@@ -420,6 +424,28 @@ func (r Routing) listCredentials() http.Handler {
...
+	return httptransport.NewServer(
+		endpoint.Chain(
+			middleware.TokenVerifier(r.tokenVerifiers),
+			middleware.UserSaver(r.userProvider),
+		)(provider.GetGCPMachineTypesEndpoint(r.credentialManager)),
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: can you change for `GCPSizeEndpoint`

**Review by zreigz — CHANGES_REQUESTED** [review]
rest looks good

**Comment by maciaszczykm** [comment]
Added endpoint for listing disk types.

Example output from `GET http://127.0.0.1:8080/api/v1/providers/gcp/disktypes`:

```json
[
    {
        "name": "pd-ssd",
        "description": "SSD Persistent Disk"
    },
    {
        "name": "pd-standard",
        "description": "Standard Persistent Disk"
    }
]
```

File: api/pkg/api/v1/types.go (hunk 2 above)
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go:149]: The description is `string` that can be useful for Dashboard. I will add the `vCPUs` raw data too.

File: api/pkg/api/v1/types.go [hunk 5]
```diff
@@ -127,6 +127,30 @@ type Datacenter struct {
...
+type GCPDiskTypeList []GCPDiskType
+
+// GCPDiskType represents a object of GCP disk type.
+// swagger:model GCPDiskType
+type GCPDiskType struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+}
+
+// GCPMachineTypeList represents an array of GCP machine types.
```
Comment (zreigz) [inline api/pkg/api/v1/types.go]: incorrect name

File: api/pkg/api/v1/types.go [hunk 6]
```diff
@@ -127,6 +127,30 @@ type Datacenter struct {
...
+// swagger:model GCPMachineTypeList
+type GCPMachineSizeList []GCPMachineSize
+
+// GCPMachineType represents a object of GCP machine type.
```
Comment (zreigz) [inline api/pkg/api/v1/types.go]: incorrect name

File: api/pkg/handler/routes_v1.go [hunk 7]
```diff
@@ -420,6 +428,50 @@ func (r Routing) listCredentials() http.Handler {
...
+// swagger:route GET /api/v1/providers/gcp/machineTypes gcp listGCPMachineTypes
+//
+// Lists machine types from GCP
+//
+//     Produces:
+//     - application/json
+//
+//     Responses:
+//       default: errorResponse
+//       200: GCPMachineTypeList
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: change type for `GCPMachineSize`

File: api/pkg/handler/routes_v1.go [hunk 8]
```diff
@@ -43,6 +43,14 @@ func (r Routing) RegisterV1(mux *mux.Router, metrics common.ServerMetrics) {
 	//
 	// Defines a set of HTTP endpoint for interacting with
 	// various cloud providers
+	mux.Methods(http.MethodGet).
+		Path("/providers/gcp/disktypes").
+		Handler(r.listGCPDiskTypes())
+
+	mux.Methods(http.MethodGet).
+		Path("/providers/gcp/machinetypes").
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: change path for `/providers/gcp/sizes`

File: api/pkg/handler/routes_v1.go [hunk 9]
```diff
@@ -43,6 +43,14 @@ func (r Routing) RegisterV1(mux *mux.Router, metrics common.ServerMetrics) {
...
+		Handler(r.listGCPMachineTypes())
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: change for `listGCPSizes`

File: api/pkg/handler/v1/provider/gcp.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+// GCPMachineTypesReq represent a request for GCP machine or disk types.
+type GCPMachineTypesReq struct {
+	ServiceAccount string
+	Zone           string
+	Credential     string
+}
+
+func DecodeGCPTypesReqReq(c context.Context, r *http.Request) (interface{}, error) {
```
Comment (zreigz) [inline api/pkg/handler/v1/provider/gcp.go]: `DecodeGCPTypesReqReq` to `DecodeGCPTypesReq`

File: api/pkg/handler/routes_v1.go [hunk 11]
```diff
@@ -420,6 +428,50 @@ func (r Routing) listCredentials() http.Handler {
...
+			middleware.TokenVerifier(r.tokenVerifiers),
+			middleware.UserSaver(r.userProvider),
+		)(provider.GCPDiskTypesEndpoint(r.credentialManager)),
+		provider.DecodeGCPTypesReqReq,
+		encodeJSON,
+		r.defaultServerOptions()...,
+	)
+}
+
+// swagger:route GET /api/v1/providers/gcp/machineTypes gcp listGCPMachineTypes
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: should be `/api/v1/providers/gcp/machinetypes` but please change for `/api/v1/providers/gcp/sizes`

**Review by zreigz — CHANGES_REQUESTED** [review]

**Comment by zreigz** [comment]
the reason for the failures:
- rebuild swagger
- rebase code

Please add unit tests for those endpoints

**Comment by maciaszczykm** [comment]
I have done some research on the tests. I have two options:

- We can export filtering based on memory (< 2 GB) or name (no `local-ssd`) to separate methods and test them as they will contain actual logic inside. There is no point in testing the connection. The valuable test would be the one testing our `/api/v1/providers/gcp/sizes` route, but I cannot find fake client for it (https://github.com/googleapis/google-api-go-client and https://cloud.google.com/compute/docs/api/libraries don't mention that).
- We can skip testing as most of the logic is connected to GCP API and establishing a connection with it. Testing private methods is not too good anyways.

**Comment by maciaszczykm** [comment]
/hold cancel

@zreigz PTAL

**Comment by zreigz** [comment]
@maciaszczykm let's skip tests right now

**Comment by maciaszczykm** [comment]
@zreigz @nikhita PTAL then.

File: api/pkg/handler/routes_v1.go [hunk 12]
```diff
@@ -420,6 +428,50 @@ func (r Routing) listCredentials() http.Handler {
 	)
 }
 
+// swagger:route GET /api/v1/providers/gcp/diskTypes gcp listGCPDiskTypes
```
Comment (zreigz) [inline api/pkg/handler/routes_v1.go]: wrong path, should be `/api/v1/providers/gcp/disktypes`

**Review by zreigz — CHANGES_REQUESTED** [review]
rest look good

**Comment by zreigz** [comment]
@maciaszczykm thanks for this PR

**Review by zreigz — APPROVED** [review]

**Pull request merged**